designs/kwok-provider.md

```diff
@@ -0,0 +1,51 @@
+
+```md
+1. Instance Type Name: (family)-(size)
+    a. size: 1x, 2x, 4x, 8x, 16x, 32x, 48x, 64x, 96x, 128x, 192x, 256x
+    b. family: c (cpu), s (standard), m (memory)
+        i. c ---> 1 vCPU : 2 GiB
+        ii. s --> 1 vCPU : 4 GiB
+        iii. m -> 1 vCPU : 8 GiB
+2. Capacity Type: [on-demand, spot]
+```
+
+#### Size
+While AWS uses `medium`, `large`, `xlarge`, `2xlarge`, `4xlarge` and so on, AKS (include cpu value in name) and GCE (suffixed with cpu value) use a different naming convention, generally increasing in offering sizes by a factor of 1.5x or 2x.
+
+#### Ratios
+AWS Instance Types fall under many "Instance Families" that give a high level overview to why a user may want to use the instance type. AWS `m` instance types are labeled General Purpose, and `c` and `r` represent Compute Optimized and Memory Optimized, most easily compared by the ratio of Memory per vCPU. Defining these ratios is important to test Karpenter's bin-packing value, where applications with more resource-skewed requirements can be more cost effectively bin-packed. 
+
+These ratios seem to accurately depict a common ratios, where "standard/general" ratios for different cloud providers are seen below:
+- AWS: General     = 1 vCPU : 4 GiB
+- AKS: D Family    = 2 vCPU : 8 GiB
+- GKE: C3 Standard = 4 vCPU : 16 GiB
+
+#### Capacity Type Names
+Karpenter v1beta1 APIs use `on-demand` and `spot` as the options for `karpenter.sh/capacity-type`, reflecting how they're referenced in AWS and elsewhere. In my docs search, I've found that Cloud Providers have different names for on-demand (e.g. Regular for AKS and Standard for GKE). Since this is how Karpenter defines the capacity type labels, the KwoK CP will use the values defined in the project, regardless of how this value changes in the future.
+
+### Pricing
+The pricing for each of the instance types will be a function of the size and capacity type.
+
+EKS: Discount is [< 90%](https://aws.amazon.com/ec2/spot/pricing/#:~:text=Spot%20Instances%20are%20available%20at%20a%20discount%20of%20up%20to%2090%25%20off%20compared%20to%20On%2DDemand%20pricing.)
+AKS: Discount is [48-90%](https://azure.microsoft.com/en-us/pricing/details/virtual-machine-scale-sets/linux/)
+GKE: Discount is [60-91%](https://cloud.google.com/compute/docs/instances/create-use-spot#:~:text=Spot%20VMs%20are%20available%20at%20a%2060%2D91%25%20discount%20compared%20to%20the%20price%20of%20standard%20VMs.)
+
+We'll define the pricing functions:
+
+```markdown
+1. price_of_base_size = (#vCPU * 0.25) + (#GiB * 0.01) # Roughly calculated based on guessing pricing functions from multiple cloud providers. 
+2. on_demand_price    = size * price_of_base_size()
+3. spot_price         = on_demand_price * .7 # Some discount factor to reflect savings somewhere in the middle of the referenced CPs.
+```
```

Comment:
If we're doing abstract testing, let's plan to one day handle a free tier.

For example, a test scenario where you can have up to two on-demand `c-1x` instances for free, regardless of your other consumption, but any use beyond that costs you actual money.

So this comment is to mark that we might one day want to account for that, not a request to put it in immediately.

Reply:
I'm not quite following. Seems pretty bespoke to simulate a user trying out Karpenter, where I don't see much of the value in that.

Reply:
Sure, don't worry about this detail.